rtengine/lensmetadata.cc

```diff
@@ -0,0 +1,1035 @@
+    } else if (ca) {
+        cf = caCorrectionFactor(rout, channel);
+    }
+
+    x = cf * xc + w2;
+    y = cf * yc + h2;
+
+    if (swap_xy) {
+        std::swap(x, y);
+    }
+    x -= cx;
+    y -= cy;
+}
+
+void CenterRadiusMetadataLensCorrection::correctDistortionAndCA(double &x, double &y, int cx, int cy, int channel) const
+{
+    if (!hasDistortionCorrection() || !hasCACorrection()) {
+        return;
+    }
+
+    process(x, y, cx, cy, channel, true, true);
+}
+
+void CenterRadiusMetadataLensCorrection::correctDistortion(double &x, double &y, int cx, int cy) const
+{
+    if (!hasDistortionCorrection()) {
+        return;
+    }
+
+    process(x, y, cx, cy, 1, true, false);
+}
+
+void CenterRadiusMetadataLensCorrection::correctCA(double &x, double &y, int cx, int cy, int channel) const
+{
+    if (!hasCACorrection()) {
+        return;
+    }
+
+    process(x, y, cx, cy, channel, false, true);
+}
+
+void CenterRadiusMetadataLensCorrection::processVignetteNChannels(int width, int height, float **rawData, int channels) const
+{
+    if (!hasVignettingCorrection()) {
+        return;
+    }
+
+    for (int y = 0; y < height; ++y) {
+        for (int x = 0; x < width; ++x) {
+            double xc = x - w2;
+            double yc = y - h2;
+            double sf = vignettingCorrectionFactor(rf * std::sqrt(SQR(xc) + SQR(yc)));
+            for (int c = 0; c < channels; c++) {
+                rawData[y][x + c] *= sf;
+            }
+        }
+    }
+}
+
+void CenterRadiusMetadataLensCorrection::processVignette(int width, int height, float **rawData) const
+{
+    return processVignetteNChannels(width, height, rawData, 1);
+}
+
+void CenterRadiusMetadataLensCorrection::processVignette3Channels(int width, int height, float **rawData) const
+{
+    return processVignetteNChannels(width, height, rawData, 3);
+}
+
+/* Fuji, Sony, Olympus metadata handling and algorithms adapted from
+ * - src/iop/lens.cc
+ * - src/common/exif.cc
+ * in darktable 4.6 */
+/*
+    This file is part of darktable,
+    Copyright (C) 2019-2024 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+class SonyMetadataLensCorrection : public CenterRadiusMetadataLensCorrection
+{
+public:
+    int nc;
+    std::array<short, 16> distortion;
+    std::array<short, 16> ca_r;
+    std::array<short, 16> ca_b;
+    std::array<short, 16> vignetting;
```

Comment:
These are the max sony knots in metadata. While MAXKNOTS is the number of knots to use to compute the inverted fuji spline. So it's a concidence that they are 16. I'll move MAXKNOTS inside the fuji class since it's used only there to avoid such confusion.